bridges/modules/xcm-bridge-hub-router/src/lib.rs

```diff
@@ -67,148 +100,276 @@ pub const HARD_MESSAGE_SIZE_LIMIT: u32 = 32 * 1024;
-		/// Also used for filtering `Bridges` by `BridgedNetworkId`.
-		/// If not specified, allows all networks pass through.
-		type BridgedNetworkId: Get<Option<NetworkId>>;
-		/// Configuration for supported **bridged networks/locations** with **bridge location** and
-		/// **possible fee**. Allows to externalize better control over allowed **bridged
-		/// networks/locations**.
-		type Bridges: ExporterFor;
 		/// Checks the XCM version for the destination.
 		type DestinationVersion: GetVersion;
 
-		/// Actual message sender (`HRMP` or `DMP`) to the sibling bridge hub location.
+		/// The bridge hub may be:
+		/// - A system (sibling) bridge hub parachain (or another chain), in which case we need an
+		///   implementation for `T::ToBridgeHubSender` that sends `ExportMessage`, e.g.,
+		///   `SovereignPaidRemoteExporter`.
+		/// - The local chain, in which case we need an implementation for `T::ToBridgeHubSender`
+		///   that does not use `ExportMessage` but instead directly calls the `ExportXcm`
+		///   implementation.
+		#[pallet::no_default]
 		type ToBridgeHubSender: SendXcm;
-		/// Local XCM channel manager.
-		type LocalXcmChannelManager: XcmChannelStatusProvider;
+
+		/// Resolves a specific `BridgeId` for `dest`, used for identifying the bridge in cases of
+		/// congestion and dynamic fees. If it resolves to `None`, it means no congestion or
+		/// dynamic fees are handled for `dest`.
+		#[pallet::no_default]
+		type BridgeIdResolver: ResolveBridgeId;
+
+		/// Origin of the sibling bridge hub that is allowed to report bridge status.
+		#[pallet::no_default]
+		type BridgeHubOrigin: EnsureOriginWithArg<Self::RuntimeOrigin, BridgeIdOf<Self, I>>;
 
 		/// Additional fee that is paid for every byte of the outbound message.
+		/// See `calculate_message_size_fees` for more details.
 		type ByteFee: Get<u128>;
-		/// Asset that is used to paid bridge fee.
-		type FeeAsset: Get<AssetId>;
+		/// Asset used to pay the `ByteFee`.
+		/// If not specified, the `ByteFee` is ignored.
+		/// See `calculate_fees` for more details.
+		type FeeAsset: Get<Option<AssetId>>;
 	}
 
+	/// An alias for the `BridgeId` of configured `T::BridgeIdResolver`.
+	pub type BridgeIdOf<T, I> = <<T as Config<I>>::BridgeIdResolver as ResolveBridgeId>::BridgeId;
+
 	#[pallet::pallet]
 	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
 
 	#[pallet::hooks]
 	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
 		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
-			// if XCM channel is still congested, we don't change anything
-			if T::LocalXcmChannelManager::is_congested(&T::SiblingBridgeHubLocation::get()) {
-				return T::WeightInfo::on_initialize_when_congested()
+			let mut weight_used = Weight::zero();
+
+			// Iterate all uncongested bridges
+			let mut bridges_to_update = Vec::new();
+			let mut bridges_to_remove = Vec::new();
+			for (bridge_id, mut bridge_state) in Bridges::<T, I>::iter() {
```

Comment:
Should we only check while we still have weight? What if there are a lot of congested bridges?

Reply:
good point, even if we don't expect too much data here, I better changed `on_initialize` to `on_idle` with weight metering, fixed by: https://github.com/paritytech/polkadot-sdk/pull/6231/commits/ecf3499b25be5d681432caafa08b9254c65b1e7a